### PR TITLE
Correction of Mass Flow Conservation Crash with Uncontrolled Zone using Mixing

### LINKED ISSUE #10107: ZoneAirMassFlowConservation object can crash when unconditioned zones are included
Issue overview
--------------
A customer file showed some issues with the ZoneAirMassFlowConservation model. A defect file was created from 5ZoneAirCooled_ZoneAirMassFlowBalance by reordering the zone objects. The ZoneAirMassFlowConservation field Infiltration Balancing Zones was changed to AllZones to match the customer file. The ZoneMixing object were set to 0 flow/zone to also match the customer file. 

The first issue with the was that the one unconditioned zone was not the first zone listed in the input file. For this reason, the difference in the zone order and the order of the ZoneReOrder array caused calculations to be performed on an unconditioned zone. In this figure line 4307 continues if the zone is uncontrolled. However, when ZoneAirMassFlowConservation is used (line 4309), ZoneNum1 is not the zone acted on, ZoneNum is the correct zone number. 

image

For this reason, line 4307 was moved down to line 4309, and ZoneNum1 changed to ZoneNum.

image

This corrected the crash later in this function where zoneEquipConfig.ZoneNode = 0.

            auto &zoneNode = Node(zoneEquipConfig.ZoneNode);
            zoneNode.MassFlowRate = TotInletAirMassFlowRate;
            zoneNode.MassFlowRateMax = TotInletAirMassFlowRateMax;
            zoneNode.MassFlowRateMaxAvail = TotInletAirMassFlowRateMaxAvail;
            zoneNode.MassFlowRateMin = TotInletAirMassFlowRateMin;
            zoneNode.MassFlowRateMinAvail = TotInletAirMassFlowRateMinAvail;

The next issue is if the ZoneMixing object flow rates are set to 0 a Nan is encountered in the simulation.

image

The reason being that the ZoneMixingReceivingFr array is only set if there is a mixing flow rate. Since the user set the flow rate to 0 this variable was never initialized.

image

This issue was corrected by initializing the ZoneMixingReceivingFr array after allocation (line 3102).

image

and also initializing ZoneMixingReceivingFr each iteration (line 4836) such that a value of ZoneMixingFlowSum = x > 0, followed by the next iterations value of ZoneMixingFlowSum = 0, would update ZoneMixingReceivingFr when ZoneMixingFlowSum = 0.0.

image

After these changes the defect file runs successfully.

    ************* EnergyPlus Completed Successfully-- 1 Warning; 0 Severe Errors; Elapsed Time=00hr 00min 19.98sec

Issue and correction courtesy of Trane North America.

### Details

Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added ([5ZoneAirCooled_ZoneAirMassFlowBalance_AllZones.idf.txt](https://github.com/NREL/EnergyPlus/files/12022076/5ZoneAirCooled_ZoneAirMassFlowBalance_AllZones.idf.txt))
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by RKStrand**
Pull request overview
---------------------
 - Fixes #10107 
 - A crash was corrected that happened during the resolution of mass flow in zones with mixing.  The problem occurred when an uncontrolled zone was involved in mixing but was not the first zone in the group of zones where mixing was occurring.  The issue was discovered by Trane who also came up with a solution.  The solution was implemented and tested and shown to correct the hard crash.  Additional code improvements in this area were also made.  This work also implements a unit test that verifies some of the initialization that was required to resolve this issue.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/HeatBalanceAirManager.cc [hunk 1]
```diff
@@ -3099,6 +3099,7 @@ void GetSimpleAirModelInputs(EnergyPlusData &state, bool &ErrorsFound) // IF err
             if (ReceivingCount > 0) {
                 state.dataHeatBal->MassConservation(ZoneNum).ZoneMixingReceivingPtr.allocate(ReceivingCount);
                 state.dataHeatBal->MassConservation(ZoneNum).ZoneMixingReceivingFr.allocate(ReceivingCount);
+                state.dataHeatBal->MassConservation(ZoneNum).ZoneMixingReceivingFr = 0.0;
```
Comment (rraustad) [inline src/EnergyPlus/HeatBalanceAirManager.cc:3102]: Same as issue suggestion

File: src/EnergyPlus/HeatBalanceAirManager.cc [hunk 2]
```diff
@@ -4816,40 +4817,34 @@ void InitSimpleMixingConvectiveHeatGains(EnergyPlusData &state)
...
-            state.dataHeatBal->Mixing(Loop).DesiredAirFlowRate =
-                state.dataHeatBal->Mixing(Loop).DesignLevel *
-                ScheduleManager::GetCurrentScheduleValue(state, state.dataHeatBal->Mixing(Loop).SchedPtr);
-            if (state.dataHeatBal->Mixing(Loop).EMSSimpleMixingOn)
-                state.dataHeatBal->Mixing(Loop).DesiredAirFlowRate = state.dataHeatBal->Mixing(Loop).EMSimpleMixingFlowRate;
-            state.dataHeatBal->Mixing(Loop).DesiredAirFlowRateSaved = state.dataHeatBal->Mixing(Loop).DesiredAirFlowRate;
+        for (auto &thisMixing : state.dataHeatBal->Mixing) {
+            thisMixing.DesiredAirFlowRate = thisMixing.DesignLevel * ScheduleManager::GetCurrentScheduleValue(state, thisMixing.SchedPtr);
+            if (thisMixing.EMSSimpleMixingOn) thisMixing.DesiredAirFlowRate = thisMixing.EMSimpleMixingFlowRate;
+            thisMixing.DesiredAirFlowRateSaved = thisMixing.DesiredAirFlowRate;
```
Comment (rraustad) [inline src/EnergyPlus/HeatBalanceAirManager.cc:4823]: No change except to convert to object loop

File: src/EnergyPlus/HeatBalanceAirManager.cc [hunk 3]
```diff
@@ -4816,40 +4817,34 @@ void InitSimpleMixingConvectiveHeatGains(EnergyPlusData &state)
...
         // Process the scheduled CrossMixing for air heat balance
-        for (int Loop = 1; Loop <= state.dataHeatBal->TotCrossMixing; ++Loop) {
-            state.dataHeatBal->CrossMixing(Loop).DesiredAirFlowRate =
-                state.dataHeatBal->CrossMixing(Loop).DesignLevel *
-                ScheduleManager::GetCurrentScheduleValue(state, state.dataHeatBal->CrossMixing(Loop).SchedPtr);
-            if (state.dataHeatBal->CrossMixing(Loop).EMSSimpleMixingOn)
-                state.dataHeatBal->CrossMixing(Loop).DesiredAirFlowRate = state.dataHeatBal->CrossMixing(Loop).EMSimpleMixingFlowRate;
+        for (auto &thisCrossMix : state.dataHeatBal->CrossMixing) {
+            thisCrossMix.DesiredAirFlowRate = thisCrossMix.DesignLevel * ScheduleManager::GetCurrentScheduleValue(state, thisCrossMix.SchedPtr);
+            if (thisCrossMix.EMSSimpleMixingOn) thisCrossMix.DesiredAirFlowRate = thisCrossMix.EMSimpleMixingFlowRate;
```
Comment (rraustad) [inline src/EnergyPlus/HeatBalanceAirManager.cc:4847]: No change except to convert to object loop

File: src/EnergyPlus/HeatBalanceAirManager.cc [hunk 4]
```diff
@@ -4860,12 +4855,12 @@ void InitSimpleMixingConvectiveHeatGains(EnergyPlusData &state)
...
-                    for (int J = 1; J <= state.dataHeatBal->RefDoorMixing(NZ).NumRefDoorConnections; ++J) {
-                        state.dataHeatBal->RefDoorMixing(NZ).VolRefDoorFlowRate(J) = 0.0;
-                        if (state.dataHeatBal->RefDoorMixing(NZ).EMSRefDoorMixingOn(J))
-                            state.dataHeatBal->RefDoorMixing(NZ).VolRefDoorFlowRate(J) = state.dataHeatBal->RefDoorMixing(NZ).EMSRefDoorFlowRate(J);
+                auto &thisRefDoor = state.dataHeatBal->RefDoorMixing(NZ);
+                if (!thisRefDoor.RefDoorMixFlag) continue;
+                if (thisRefDoor.ZonePtr == NZ) {
+                    for (int J = 1; J <= thisRefDoor.NumRefDoorConnections; ++J) {
+                        thisRefDoor.VolRefDoorFlowRate(J) = 0.0;
+                        if (thisRefDoor.EMSRefDoorMixingOn(J)) thisRefDoor.VolRefDoorFlowRate(J) = thisRefDoor.EMSRefDoorFlowRate(J);
```
Comment (rraustad) [inline src/EnergyPlus/HeatBalanceAirManager.cc:4863]: No change except to convert to reference

File: src/EnergyPlus/ZoneEquipmentManager.cc [hunk 5]
```diff
@@ -4501,9 +4501,9 @@ void CalcZoneMassBalance(EnergyPlusData &state, bool const FirstHVACIteration)
         BuildingZoneReturnFlow = 0.0;
 
         for (int ZoneNum1 = 1; ZoneNum1 <= state.dataGlobal->NumOfZones; ++ZoneNum1) {
-            if (!state.dataZoneEquip->ZoneEquipConfig(ZoneNum1).IsControlled) continue;
             int ZoneNum = ZoneNum1;
             if (state.dataHeatBal->ZoneAirMassFlow.EnforceZoneMassBalance) ZoneNum = state.dataHeatBalFanSys->ZoneReOrder(ZoneNum1);
+            if (!state.dataZoneEquip->ZoneEquipConfig(ZoneNum).IsControlled) continue;
```
Comment (rraustad) [inline src/EnergyPlus/ZoneEquipmentManager.cc:4506]: Same as issue suggestion

File: src/EnergyPlus/HeatBalanceAirManager.cc [hunk 6]
```diff
@@ -4816,40 +4817,34 @@ void InitSimpleMixingConvectiveHeatGains(EnergyPlusData &state)
...
                 for (int Loop = 1; Loop <= NumOfMixingObjects; ++Loop) {
-                    ZoneMixingFlowSum = ZoneMixingFlowSum + state.dataHeatBal->Mixing(Loop).DesignLevel;
+                    ZoneMixingFlowSum += state.dataHeatBal->Mixing(Loop).DesignLevel;
+                    massConservZone.ZoneMixingReceivingFr(Loop) = 0.0;
                 }
                 if (ZoneMixingFlowSum > 0.0) {
                     for (int Loop = 1; Loop <= NumOfMixingObjects; ++Loop) {
-                        state.dataHeatBal->MassConservation(ZoneNum).ZoneMixingReceivingFr(Loop) =
-                            state.dataHeatBal->Mixing(Loop).DesignLevel / ZoneMixingFlowSum;
+                        massConservZone.ZoneMixingReceivingFr(Loop) = state.dataHeatBal->Mixing(Loop).DesignLevel / ZoneMixingFlowSum;
```
Comment (rraustad) [inline src/EnergyPlus/HeatBalanceAirManager.cc:4838]: So, with other code changes here that subtly target the code, that leaves this code change (4829-4838) that switches from a zone loop to a mixing object loop. This change does not appear different from the original intent except that now the ZoneMixingReceivingFr array is zero'd before checking whether `ZoneMixingFlowSum > 0.0` at line 4836. This too is an issue description suggestion.

    state.dataHeatBal->MassConservation.allocate(state.dataGlobal->NumOfZones);

File: tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc [hunk 7]
```diff
@@ -860,4 +860,62 @@ TEST_F(EnergyPlusFixture, HeatBalanceAirManager_GetMixingAndCrossMixing)
...
+    state->dataHeatBal->Mixing(2).DesignLevel = 300.0;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1) = -9999.9;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2) = -9999.9;
+    expectedResult1 = 0.25;
+    expectedResult2 = 0.75;
+    HeatBalanceAirManager::InitSimpleMixingConvectiveHeatGains(*state);
+    EXPECT_NEAR(expectedResult1, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1), allowedTolerance);
+    EXPECT_NEAR(expectedResult2, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2), allowedTolerance);
+}
+
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc:920]: I'll need to think about this unit test when I have more time.

**Review by rraustad — COMMENTED** [review]
This looks good @RKStrand. From what I can see only the original suggestions were added and nothing else has changed. Still need to test a few things (i.e., try to break it) and go over the unit test.

**Comment by rraustad** [comment]
There are no diffs, which is expected, since this change fixes a corner case.

File: tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc [hunk 8]
```diff
@@ -860,4 +860,62 @@ TEST_F(EnergyPlusFixture, HeatBalanceAirManager_GetMixingAndCrossMixing)
...
+    // Test 1: not air flow flag--don't do anything
+    state->dataHeatBal->AirFlowFlag = false;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1) = -9999.9;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2) = -9999.9;
+    expectedResult1 = -9999.9;
+    expectedResult2 = -9999.9;
+    HeatBalanceAirManager::InitSimpleMixingConvectiveHeatGains(*state);
+    EXPECT_NEAR(expectedResult1, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1), allowedTolerance);
+    EXPECT_NEAR(expectedResult2, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2), allowedTolerance);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc:894]: If AirFlowFlag = false and EnforceZoneMassBalance = true, these will not get initialized to 0.

File: tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc [hunk 9]
```diff
@@ -860,4 +860,62 @@ TEST_F(EnergyPlusFixture, HeatBalanceAirManager_GetMixingAndCrossMixing)
...
+    state->dataHeatBal->AirFlowFlag = true;
+    state->dataHeatBal->Mixing(1).DesignLevel = 0.0;
+    state->dataHeatBal->Mixing(2).DesignLevel = 0.0;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1) = -9999.9;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2) = -9999.9;
+    expectedResult1 = 0.0;
+    expectedResult2 = 0.0;
+    HeatBalanceAirManager::InitSimpleMixingConvectiveHeatGains(*state);
+    EXPECT_NEAR(expectedResult1, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1), allowedTolerance);
+    EXPECT_NEAR(expectedResult2, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2), allowedTolerance);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc:906]: And with AirFlowFlag = true and EnforceZoneMassBalance = true, these will get initialized to 0.

File: tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc [hunk 10]
```diff
@@ -860,4 +860,62 @@ TEST_F(EnergyPlusFixture, HeatBalanceAirManager_GetMixingAndCrossMixing)
...
+    state->dataHeatBal->AirFlowFlag = true;
+    state->dataHeatBal->Mixing(1).DesignLevel = 100.0;
+    state->dataHeatBal->Mixing(2).DesignLevel = 300.0;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1) = -9999.9;
+    state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2) = -9999.9;
+    expectedResult1 = 0.25;
+    expectedResult2 = 0.75;
+    HeatBalanceAirManager::InitSimpleMixingConvectiveHeatGains(*state);
+    EXPECT_NEAR(expectedResult1, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(1), allowedTolerance);
+    EXPECT_NEAR(expectedResult2, state->dataHeatBal->MassConservation(1).ZoneMixingReceivingFr(2), allowedTolerance);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/HeatBalanceAirManager.unit.cc:918]: Here, where there are valid values for DesignLevel, of which the sum is 400, ZoneMixingReceivingFr(1) = 0.25 (100/400) and ZoneMixingReceivingFr(2) = 0.75 (300/400). So unit test looks good.

**Review by rraustad — COMMENTED** [review]
Looked more closely at unit test and this looks good. Will test defect file now.

**Comment by rraustad** [comment]
Pull develop and unit tests run successfully.
Defect file runs successfully in this branch using release and debug.
Defect file runs successfully in develop using release and crashes in debug:

image

image

**Comment by Myoldmopar** [comment]
That's awesome, thanks @rraustad for the testing!

**Comment by rraustad** [comment]
Original defect file with mixing set to 0, top 6 plots. And bottom 6 plots show results with mixing set to non-zero values (e.g., 0,051, 0.052, 0.053, and 0.054 for mixing objects in zone 5 connected to zones 1 - 4.

image

I am assuming zone 5 is the receiving zone from each of the other zones so this result looks correct.

  ZoneMixing,
    SPACE5-1 Zone Mixing1,   !- Name
    SPACE5-1,                !- Zone or Space Name
    ZoneMixing-SCH,          !- Schedule Name
    Flow/Zone,               !- Design Flow Rate Calculation Method
    0.051,                    !- Design Flow Rate {m3/s}
    ,                        !- Flow Rate per Floor Area {m3/s-m2}
    ,                        !- Flow Rate per Person {m3/s-person}
    ,                        !- Air Changes per Hour {1/hr}
    SPACE1-1,                !- Source Zone or Space Name
    0.0,                     !- Delta Temperature {deltaC}
    ,                        !- Delta Temperature Schedule Name
    ,                        !- Minimum Receiving Temperature Schedule Name
    ,                        !- Maximum Receiving Temperature Schedule Name
    ,                        !- Minimum Source Temperature Schedule Name
    ,                        !- Maximum Source Temperature Schedule Name
    ,                        !- Minimum Outdoor Temperature Schedule Name
    ;                        !- Maximum Outdoor Temperature Schedule Name

  ZoneMixing,
    SPACE5-1 Zone Mixing2,   !- Name
    SPACE5-1,                !- Zone or Space Name
    ZoneMixing-SCH,          !- Schedule Name
    Flow/Zone,               !- Design Flow Rate Calculation Method
    0.052,                    !- Design Flow Rate {m3/s}
    ,                        !- Flow Rate per Floor Area {m3/s-m2}
    ,                        !- Flow Rate per Person {m3/s-person}
    ,                        !- Air Changes per Hour {1/hr}
    SPACE2-1,                !- Source Zone or Space Name
    0.0,                     !- Delta Temperature {deltaC}
    ,                        !- Delta Temperature Schedule Name
    ,                        !- Minimum Receiving Temperature Schedule Name
    ,                        !- Maximum Receiving Temperature Schedule Name
    ,                        !- Minimum Source Temperature Schedule Name
    ,                        !- Maximum Source Temperature Schedule Name
    ,                        !- Minimum Outdoor Temperature Schedule Name
    ;                        !- Maximum Outdoor Temperature Schedule Name

  ZoneMixing,
    SPACE5-1 Zone Mixing3,   !- Name
    SPACE5-1,                !- Zone or Space Name
    ZoneMixing-SCH,          !- Schedule Name
    Flow/Zone,               !- Design Flow Rate Calculation Method
    0.053,                    !- Design Flow Rate {m3/s}
    ,                        !- Flow Rate per Floor Area {m3/s-m2}
    ,                        !- Flow Rate per Person {m3/s-person}
    ,                        !- Air Changes per Hour {1/hr}
    SPACE3-1,                !- Source Zone or Space Name
    0.0,                     !- Delta Temperature {deltaC}
    ,                        !- Delta Temperature Schedule Name
    ,                        !- Minimum Receiving Temperature Schedule Name
    ,                        !- Maximum Receiving Temperature Schedule Name
    ,                        !- Minimum Source Temperature Schedule Name
    ,                        !- Maximum Source Temperature Schedule Name
    ,                        !- Minimum Outdoor Temperature Schedule Name
    ;                        !- Maximum Outdoor Temperature Schedule Name

  ZoneMixing,
    SPACE5-1 Zone Mixing4,   !- Name
    SPACE5-1,                !- Zone or Space Name
    ZoneMixing-SCH,          !- Schedule Name
    Flow/Zone,               !- Design Flow Rate Calculation Method
    0.054,                    !- Design Flow Rate {m3/s}
    ,                        !- Flow Rate per Floor Area {m3/s-m2}
    ,                        !- Flow Rate per Person {m3/s-person}
    ,                        !- Air Changes per Hour {1/hr}
    SPACE4-1,                !- Source Zone or Space Name
    0.0,                     !- Delta Temperature {deltaC}
    ,                        !- Delta Temperature Schedule Name
    ,                        !- Minimum Receiving Temperature Schedule Name
    ,                        !- Maximum Receiving Temperature Schedule Name
    ,                        !- Minimum Source Temperature Schedule Name
    ,                        !- Maximum Source Temperature Schedule Name
    ,                        !- Minimum Outdoor Temperature Schedule Name
    ;                        !- Maximum Outdoor Temperature Schedule Name

**Review by rraustad — APPROVED** [review]
@Myoldmopar I think this is ready for merge.

File: src/EnergyPlus/HeatBalanceAirManager.cc (hunk 2 above)
Comment (Myoldmopar) [reply src/EnergyPlus/HeatBalanceAirManager.cc:4823]: This is very nice.

**Review by Myoldmopar — APPROVED** [review]
Simple set of changes with some cleanup. :+1:

**Comment by Myoldmopar** [comment]
Yep, all good.  Pulled develop and it's passing just fine.  Thanks @RKStrand and @rraustad !
